Atlas plugin

sajit opened the pull request:
## What does this PR change?
* Introduces method to create cost explorer query

## Does this PR relate to any other PRs?
*  No

## How will this PR impact users?
* Not sure 

## Does this PR address any GitHub or Zendesk issues?
*  No 

## How was this PR tested?
* Unit tets

## Does this PR require changes to documentation?
*

sajit commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,99 @@
+
+		EndDate:       "2024-07-01",
+		StartDate:     "2023-12-01",
+		Organizations: []string{"5c06ddb279358e17bfcfca49"},
```
For testing

ameijer commented:
Amazing, Sajit! I created an issue for tracking here: https://github.com/opencost/opencost-plugins/issues/24

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,99 @@
+
+	payloadJson, marshallingError := json.Marshal(payload)
+	if marshallingError != nil {
+		return "error", marshallingError
```
golang error handling is one of the most interesting and different parts of the language. Even after 5 years of it being my daily driver, I can't decide if I like it better or not than Java. 

Basically, go never throws exceptions. if you do a nil access, it will panic with a segfault like C, but you can't use exception handling to deal with unexpected errors. you will see error handles like this sprinkled throughout all go code. 

We have found the following approach works:
1. when returning an error, your "good" value should be nil if a pointer, or an empty value if not a pointer.
2. log the error in context (import the logging package from opencost, that will handle logger setup for you, I can point you to examples)
3. wrap the error before you return it. As the error gets passed up the stack, you have nested wrapped messages which function similar to the stacktrace in a java exception. 

Here's how I would handle this error in this case:

```
if marshallingError != nil {
   msg := fmt.Sprintf("createCostExplorerQueryToken: error marshaling JSON payload: %v", err)
   log.Error(msg)
   return "", fmt.Errorf(msg)
}
```

sajit replied:
Will take point on this.

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,99 @@
+	}
+	defer response.Body.Close()
+
+	fmt.Println("response Status:", response.Status)
```
I'd go ahead and move all this to a debug or trace level logger. This will be useful for ongoing maintenance

ameijer commented on mongodb-atlas/mongodbatlasplugin/mongodb-atlas-config.go:
```diff
@@ -0,0 +1,6 @@
+package mongodbatlas
+
+type MongoDBAtlasConfig struct {
+	PublicKey  string
```
I'd go ahead and get this reading off the file system during the boot sequence. Check out how we do it in `getDatadogConfig` and `getConfigFilePath()` in that plugin. This config will grow quickly as this matures and you will be glad to get it in a file ASAP before you get trapped in env var hell world

ameijer reviewed (commented):
regarding project structure - I think you are actually pretty close. The `main` package is conventionally where the main() method is... you will always have a main method, each plugin is its own standalone go program. 

I regret naming the plugin implementation package `datadogplugin`, i would change the mongodbatlasplugin package to just be named `plugin`... less typing to use that stuff, haha.

I'd also set up a test package that can hold an integration testing harness for this

kwombach12 commented:
@sajit Thanks so much for your contribution! We are super excited to work with you on this. Please don't hesitate to reach out if you have any issues or questions.

sajit commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,99 @@
+// TODO move these types to mongodbatlasplugin package
+type CreateCostExplorerQueryPayload struct {
+	Clusters              []string `json:"clusters"`
+	EndDate               string   `json:"endDate"`
```
Ive modified the function itself to take time.Time.  So its more strict. Internally we are creating this struct.  when I tried changing EndDate to time.Time issue I was running into was that the string was of the format yyyy-mm-dd+timestamp and mongo atlas didn't like it. Instead of figuring out how to modify the json.Marshal to marshal a time type to yyyy-mm-dd without the timestamp I just do a parse in the function itself.

sajit commented:
@ameijer  @kwombach12  Thanks for the feedback. Kindly take another look. Ive addressed most of the comments / deleted unused code. A few things to note. In main_test.go, all but 1 test are unit tests.  
For,  `TestMain` test in `mongodb-atlas/cmd/main_test.go` to run the execution environment , needs to have some env variables set . (its documented in the comments) . This is a true integration test - ie it will make a real call to MongoDB atlas.

sajit commented:
Also could I get some help understanding whats happening with DCO - I believe I have used the correct commit message template.  thanks in advance

sajit commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+// pass list of orgs , start date, end date
+func createCostExplorerQueryToken(org string, startDate time.Time, endDate time.Time,
```
Changed startDate, endDate to be time.Time

ameijer commented:
wow the DCO thing is weird. Have you set up GPG keys? is it related to that? 

If all else fails, you can always just copy the files into a new branch

sajit commented:
@ameijer pretty silly apparently its because DCO requires your first name and last name and my github was just my first name. lol. Anyways I think its passing DCO check now. Should be good to be reviewed.

ameijer commented on mongodb-atlas/plugin/costexplorerquery.go:
```diff
@@ -0,0 +1,29 @@
+	Service          string  `json:"service"`
+	UsageAmount      float32 `json:"usageAmount"`
+	UsageDate        string  `json:"usageDate"`
+	//"invoiceId":"66d7254246a21a41036ff315","organizationId":"66d7254246a21a41036ff2e9","organizationName":"Kubecost","service":"Clusters","usageAmount":51.19,"usageDate":"2024-09-01"}
```
can you redact the IDs here?

ameijer commented on mongodb-atlas/plugin/costexplorerquery.go:
```diff
@@ -0,0 +1,29 @@
+package plugin
+
+type CreateCostExplorerQueryPayload struct {
```
very nice

ameijer commented on mongodb-atlas/config/atlasconfig_test.go:
```diff
@@ -0,0 +1,76 @@
+	// 	if config.LogLevel != "debug" {
+	// 		t.Errorf("expected log level to be 'debug', but got: %s", config.LogLevel)
+	// 	}
+	// })
```
why commented?

ameijer commented on .gitmessage.txt:
```diff
@@ -0,0 +1,2 @@
+#Commit Message template
```
can you remove these files?

ameijer commented on .vscode/launch.json:
```diff
@@ -0,0 +1,17 @@
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": [
```
I gitignored this and removed it in https://github.com/sajit/opencost-plugins/pull/4

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+const costExplorerQueryFmt = "https://cloud.mongodb.com/api/atlas/v2/orgs/%s/billing/costExplorer/usage/%s"
+
+func main() {
+	fmt.Println("Initializing Mongo plugin")
```
changed to be a debug log in https://github.com/sajit/opencost-plugins/pull/4, the test harness exposed a bug when the plugin prints to stdout

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+func (a *AtlasCostSource) getAtlasCostsForWindow(win *opencost.Window) (*pb.CustomCostResponse, error) {
+
+	// get the token
+	token, err := CreateCostExplorerQueryToken(a.orgID, *win.Start(), *win.End(), a.atlasClient)
```
so, the test harness exposed an issue where we can't use anything but the start of the month. WHat you will want to do here is round the window up to months. you can I think call getMonthlyWindow on this window object, however, I see it is not exported from opencost so you might just have to copy paste the code:

```
func (w Window) getMonthlyWindow() Window {
	origStart := w.Start()
	origEnd := w.End()
	// round the start and end windows to the calendar month start and ends, respectively
	roundedStart := time.Date(origStart.Year(), origStart.Month(), 1, 0, 0, 0, 0, origStart.Location())
	roundedEnd := time.Date(origEnd.Year(), origEnd.Month()+1, 1, 0, 0, 0, 0, origEnd.Location())
	// edge case - if user has exactly specified first instant of new month, does not need rounding
	if origEnd.Day() == 1 && origEnd.Hour() == 0 && origEnd.Minute() == 0 && origEnd.Second() == 0 {
		roundedEnd = *origEnd
	}
	return NewClosedWindow(roundedStart, roundedEnd)
}
```

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+	resp := pb.CustomCostResponse{
+		Metadata:   map[string]string{"api_client_version": "v1"},
+		CostSource: "data_storage",
+		Domain:     "mongodb-atlas",
```
const for this

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+
+	resp := pb.CustomCostResponse{
+		Metadata:   map[string]string{"api_client_version": "v1"},
+		CostSource: "data_storage",
```
I would make this be 'Data Storage', so it displays better?

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+	for count := 1; count < 2 && statusCode == http.StatusProcessing; count++ {
+		// Sleep for 5 seconds before the next request
+		time.Sleep(5 * time.Second)
+		response, _ := client.Do(request)
```
check error and log at least? Any way that you can sleep _after_ the request is sent? just to avoid always a delay

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+		msg := "timeout waiting for response"
+		return nil, fmt.Errorf(msg)
+	}
+	if error != nil {
```
i would be very very very careful about using error as a variable name. It is a protected name in many situations in go since it is the type of the error values. 95% of the time, we use err as the name to avoid confusion

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+			Id:             invoice.InvoiceId,
+			AccountName:    invoice.OrganizationName,
+			ChargeCategory: invoice.Service,
+			BilledCost:     invoice.UsageAmount,
```
I would instead use ListCost to store this in. We used billedCost when hitting invoice APIs, not as often for usage apis like we are hitting now

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+	for _, invoice := range costResponse.UsageDetails {
+		// Create a new pb.CustomCost for each Invoice
+		customCost := &pb.CustomCost{
+			Id:             invoice.InvoiceId,
```
this ID is actually something you assign. You can just do UUID.NewUUIDV4String() from one of the many UUID packages out there.

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+func CreateCostExplorerQueryToken(org string, startDate time.Time, endDate time.Time,
+	client HTTPClient) (string, error) {
+	// Define the layout for the desired format
+	layout := "2006-01-02"
```
const

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+
+	payload := atlasplugin.CreateCostExplorerQueryPayload{
+
+		EndDate:       endDateString,
```
rm space above this

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+	// Iterate over the UsageDetails in CostResponse
+	for _, invoice := range costResponse.UsageDetails {
+		// Create a new pb.CustomCost for each Invoice
+		customCost := &pb.CustomCost{
```
can you set UsageUnit as USD on this too?

ameijer replied:
on the metadata field, can you set an interpolated: true key/value?

ameijer replied:
another additional field that is super important is the ProviderID. We can sort of synthesize one. Maybe the providerID can be `<orgid>/<invoiceid>/<service>`

ameijer commented on mongodb-atlas/cmd/main.go:
```diff
@@ -0,0 +1,253 @@
+		customCost := &pb.CustomCost{
+			Id:             invoice.InvoiceId,
+			AccountName:    invoice.OrganizationName,
+			ChargeCategory: invoice.Service,
```
I would actually set ResourceName to the invoice.Service

ameijer reviewed (commented):
exciting stuff!

ameijer commented:
obsoleted by https://github.com/opencost/opencost-plugins/pull/41
